create_disk: Use bootupd for BIOS too

cgwalters opened the pull request:
build: Fast track bootupd

Because waiting an entire week for no one to add karma is just
the worst...

---

create_disk: Use bootupd for BIOS too

This uses the new https://github.com/coreos/bootupd/ 0.2.11 release.

---

cgwalters commented on src/create_disk.sh:
```diff
@@ -477,22 +481,10 @@ generate_gpgkeys() {
-        # https://github.com/coreos/fedora-coreos-tracker/issues/32
-        # Install BIOS/PReP bootloader using the target system's grub2-install,
-        # see https://github.com/coreos/coreos-assembler/issues/3156
-        chroot_run /sbin/grub2-install \
```
Once this merges, another little bit of shell script dies and becomes Rust...

dustymabe commented on src/create_disk.sh:
```diff
@@ -477,22 +481,10 @@ generate_gpgkeys() {
 case "$arch" in
 x86_64)
     # UEFI
```
```suggestion
    # bootupd does UEFI and BIOS now
```

or I guess we could just delete the comment

dustymabe commented on build.sh:
```diff
@@ -71,7 +71,7 @@ install_rpms() {
     # Add fast-tracked packages here.  We don't want to wait on bodhi for rpm-ostree
     # as we want to enable fast iteration there.
-    yum -y --enablerepo=updates-testing upgrade rpm-ostree
+    yum -y --enablerepo=updates-testing upgrade rpm-ostree bootupd
```
honestly it feels like we should be running the bootupd from inside the installed system and not the one from COSA. We had a [similar problem](https://github.com/coreos/coreos-assembler/issues/3156) with grub2 which is why we were using `chroot_run` before for the `grub2-install`

cgwalters replied:
Yes, good catch!  This would be a notable regression for RHCOS.  Um... @HuijingHei you and I discussed this I'm pretty sure (but it was way back in February)...not finding it in https://github.com/coreos/bootupd/pull/420 though?

I think you might be right though and our conclusion was that bootupd itself shouldn't be doing a chroot, but should be run in a (privileged) container.  Will look at this.

cgwalters replied:
OK, switched to wrap in `chroot_run`! 

Also of note: this is how bootc does it, because in `bootc install` land, the *container installs itself* so you're already running in the target system.

HuijingHei replied:
> I think you might be right though and our conclusion was that bootupd itself shouldn't be doing a chroot, but should be run in a (privileged) container. Will look at this.

Agree with you.

dustymabe commented on src/create_disk.sh:
```diff
@@ -396,11 +396,15 @@ cat > $rootfs/.coreos-aleph-version.json << EOF
 EOF
 
-install_uefi() {
+install_via_bootupd() {
```
in case this causes regressions maybe we should make this a image.yaml opt-in knob and slowly roll it out?

i.e. I think I'd be happy to enable it in `next` now and then let that flow into testing and stable along with the f39 rebase in october.

dustymabe replied:
we have a few other image.yaml knobs for the AMI changes that we're already going to shepherd through with the f39 update so we could just manage them together probably.

cgwalters replied:
> in case this causes regressions maybe we should make this a image.yaml opt-in knob and slowly roll it out?


Mmmm....not opposed...if you feel slightly strongly.  Wouldn't be too hard.

OTOH, BIOS boot basically either works or it doesn't (and I've tested this now, and I think Huijing did a good bit of testing in development too).

dustymabe replied:
interesting in other takes here: cc @travier, @jlebon

dustymabe replied:
one other piece of info.. if we merge this as is we'd need to tag a COSA from before this lands and use it to build the next `stable` release because a new enough bootupd won't be in that payload IIUC. for the next `testing` and `next` we can fast-track the bootupd pending update.

cgwalters replied:
Actually the more obvious problem is it won't work at all with RHCOS for some time unless we fast track it there...so yeah we need a config knob.

dustymabe commented:
Thanks for working on this

dustymabe commented on src/create_disk.sh:
```diff
@@ -476,23 +480,10 @@ generate_gpgkeys() {
-    install_uefi
+    install_via_bootupd
     ;;
 ppc64le)
```
without ppc64le support in bootupd I think it might make our "implement this in OSBuild" story a bit more complicated. Are we considering adding ppc64le support to bootupd?

cgwalters replied:
Ah right, sorry I forgot we actually *did* add PReP support in that same PR, so I've amended the PR (untested) to also add that.

And now that we've done that, maybe we should go ahead and add the (really trivial) support for zipl...

cgwalters replied:
Oh right, zipl isn't trivial at all...lots of logic in https://github.com/coreos/coreos-installer/blob/main/src/s390x/zipl.rs

Which gets into a tricky problem in having bootupd depend on coreos-installer (or worse, bootc depend on it).

We probably need to factor that logic out into a crate.

cgwalters replied:
The more general problem is that s390x/zipl actually mainly ties together the kernel/initramfs/bootloader - which in many ways is a good thing - it's actually what we *also* want to support for a UKI-style flow, but *that* then merges the problem domain of bootloader and OS update.

dustymabe replied:
> Ah right, sorry I forgot we actually _did_ add PReP support in that same PR, so I've amended the PR (untested) to also add that.

Oh nice. I didn't catch that 👍

ravanelli reviewed (dismissed):
Thanks for adding it @cgwalters! 
Overall LGTM, waiting to see https://github.com/coreos/coreos-assembler/pull/3631#discussion_r1332193110

cgwalters commented:
Now with a config knob, off by default.  This definitely needs some careful review (whee shell script fragility) and testing (across all architectures really).

dustymabe commented:
> Now with a config knob, off by default. This definitely needs some careful review (whee shell script fragility) and testing (across all architectures really).

+1 - i'll take a closer look in the morning when I'm fresh. Once we get it close on code review we can do a COSA build in the pipeline and run it through the bump-lockfile job in the staging pipeline (tests against all architectures).

HuijingHei commented on src/create_disk.sh:
```diff
@@ -472,32 +487,50 @@ generate_gpgkeys() {
+            install_bootloader_nons390x_bootupd_epoch_0
+            ;;
+        1)
+            install_via_bootupd
```
Overall looks good to me, one minor comment, I think you meant `bootloader_bootupd_epoch_1`?

jlebon replied:
Looks like it to me too.

cgwalters commented on build.sh:
```diff
@@ -71,7 +71,7 @@ install_rpms() {
     # Add fast-tracked packages here.  We don't want to wait on bodhi for rpm-ostree
     # as we want to enable fast iteration there.
-    yum -y --enablerepo=updates-testing upgrade rpm-ostree
+    yum -y --enablerepo=updates-testing upgrade rpm-ostree bootupd
```
Yep.

jlebon commented on src/create_disk.sh:
```diff
@@ -396,11 +397,25 @@ cat > $rootfs/.coreos-aleph-version.json << EOF
+        bootupd_args+=("--device=$disk")
+    fi
+    chroot_run /usr/bin/bootupctl backend install --src-root="${deploy_root}" "${bootupd_args[@]}" "${rootfs}"
+    inject_grub_uefi
```
Shouldn't this be arch-specific? On ppc64le, we only want `install_grub_cfg`.

cgwalters commented:
Tested now on x86_64, booting BIOS and UEFI in both epoch 0 and 1.

cgwalters commented:
Note that we're also discussing moving the static grub configs into a subpackage of bootupd, xref https://github.com/coreos/bootupd/pull/536

If we did that, it'd be another epoch.  (Which, if we agree to do it, maybe we should just make "epoch 1" be that)

cgwalters commented:
Renata added this on internal chat but replicating here:  Once we do this we also want to update https://docs.fedoraproject.org/en-US/fedora-coreos/bootloader-updates/

(That said, in the context of [Sagano](https://gitlab.com/CentOS/cloud/sagano) this would actually be a Sagano doc, not a CoreOS doc...)

jlebon commented:
Interesting, CI just hit https://github.com/coreos/rpm-ostree/issues/4565. So I guess https://github.com/coreos/coreos-assembler/pull/3619 didn't do the trick.

jlebon reviewed (approved):
LGTM! (Would still be good to do some multi-arch testing but it being opt in means we can easily test it in e.g. rawhide first before propagating.)

> Note that we're also discussing moving the static grub configs into a subpackage of bootupd, xref https://github.com/coreos/bootupd/pull/536

I'm not sure that makes sense. Posted in https://github.com/coreos/bootupd/pull/536#issuecomment-1730147912.

dustymabe commented on src/create_disk.sh:
```diff
@@ -396,11 +397,34 @@ cat > $rootfs/.coreos-aleph-version.json << EOF
+    chroot_run /usr/bin/bootupctl backend install --src-root="${deploy_root}" "${bootupd_args[@]}" "${rootfs}"
+    case "${arch}" in
+        x86_64|aarch64)
+            inject_grub_uefi
```
optional: I think this would be slightly easier to follow if the `inject_grub_cfg` wasn't embedded in `inject_grub_uefi` but rather called at the same level:

```suggestion
            inject_grub_uefi
            inject_grub_cfg
```

i.e. it would make it clearer that x86_64 and aarch64 need BOTH `inject_grub_uefi` and `inject_grub_cfg`, but ppc64le only needs `inject_grub_cfg`.

dustymabe commented on src/create_disk.sh:
```diff
@@ -396,11 +397,34 @@ cat > $rootfs/.coreos-aleph-version.json << EOF
+    # For background on bootupd, see https://github.com/coreos/fedora-coreos-tracker/issues/510
     # Unshare mount ns to work around https://github.com/coreos/bootupd/issues/367
     unshare -m /usr/bin/bootupctl backend install --src-root="${deploy_root}" "${rootfs}"
+    inject_grub_uefi
```
if we do https://github.com/coreos/coreos-assembler/pull/3631/files#r1334584042 then we'd need to add a call here:

```suggestion
    inject_grub_uefi
    inject_grub_cfg
```

dustymabe commented:
Let me know what you think of the optional comments. Otherwise everything LGTM. I can do a multi-arch build of COSA and run this through the staging bump-lockfile job with both epoch 0 and 1 to see how the architectures respond to it if you like.

cgwalters commented:
Sure, SGTM (honestly feel free to force push cleanups here if you like).

That said I think I'd actually want to wait and make "epoch 1" include the grub.cfg rework, so will release a new bootupd soon.

HuijingHei commented:
Do testing with default `bootupd-epoch=0`, build cosa and image, start with `uefi` firmware, check 2 service are failed.
```

$ cosa run -m 4096 --qemu-firmware uefi 
[systemd]
Failed Units: 2
  var-mnt-workdir.mount
  var-mnt-workdir\x2dtmp.mount

[core@cosa-devsh ~]$ systemctl status var-mnt-workdir.mount
× var-mnt-workdir.mount - /var/mnt/workdir
     Loaded: loaded (/etc/systemd/system/var-mnt-workdir.mount; enabled; preset: enabled)
     Active: failed (Result: exit-code) since Mon 2023-09-25 10:23:54 UTC; 2min 17s ago
      Where: /var/mnt/workdir
       What: /var/mnt/workdir
        CPU: 9ms

Sep 25 10:23:54 localhost systemd[1]: Mounting var-mnt-workdir.mount - /var/mnt/workdir...
Sep 25 10:23:54 localhost mount[1442]: mount: /var/mnt/workdir: can't read superblock on /var/mnt/workdir.
Sep 25 10:23:54 localhost mount[1442]:        dmesg(1) may have more information after failed mount system call.
Sep 25 10:23:54 localhost systemd[1]: var-mnt-workdir.mount: Mount process exited, code=exited, status=32/n/a
Sep 25 10:23:54 localhost systemd[1]: var-mnt-workdir.mount: Failed with result 'exit-code'.
Sep 25 10:23:54 localhost systemd[1]: Failed to mount var-mnt-workdir.mount - /var/mnt/workdir.

$ dmesg | grep virtio
[   19.112948] virtiofs virtio4: requests.0:id 0 is not a head!
[   19.115108] virtio-fs: virtio_fs_enqueue_req() failed -5
[   19.115114] SELinux: (dev virtiofs, type virtiofs) getxattr errno 5
...
```

cgwalters commented:
See https://github.com/coreos/bootupd/pull/543 (not totally proud of it, kind of hacky but...) anyways this is now updated to also take ownership of most of the grub config, what's left here is `platform.cfg` stuff.

I wonder if this PR would be better off actually just fully deleting the old code?  Landing now would make it easier to test locally, but only slightly.  The patch here would be way less ugly if it was 70% deletions instead.

cgwalters commented:
OK this now builds on https://github.com/coreos/bootupd/pull/543 and https://github.com/coreos/ignition/pull/1728
And I've reworked it to just completely delete the old code.  We can land this once we've updated bootupd and ignition in fcos (ugh I guess all the way to stable?) and tagged into rhcos.

cgwalters commented:
I guess one option is to basically close this PR and aim to try to cut over to osbuild in https://github.com/coreos/coreos-assembler/pull/3643 ?

dustymabe commented:
I think this PR is still useful. I wanted to get back to it and run it through some tests on multi-arch in the staging pipeline.

This has also been nice inspiration for @ravanelli who is working on a bootupd stage in osbuild.

cgwalters commented:
Right, it was useful to write this to sanity check the bootupd support.  

(Incidentally https://github.com/containers/bootc/pull/157 also landed)

dustymabe commented:
going to close this out since we are moving away from using create_disk for our disk image creation.
